### PR TITLE
PartialTransaction: replace nested Eithers by a custom ADT

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
requested by @cocreature and @SamirTalwar in #7189

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Thanks for addressing this so quickly!

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala [hunk 1]
```diff
@@ -427,10 +428,10 @@ class IdeClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClie
                     // Capture the result and exit.
                     result = results.toSeq
                 }
-              case Right(Left(x)) =>
-                throw new RuntimeException(s"Unexpected abort: $x")
-              case Left(msg) =>
-                throw new RuntimeException(msg)
+              case PartialTransaction.IncompleteTransaction(ptx) =>
+                result = Failure(new RuntimeException(s"Unexpected abort: $ptx"))
```
Comment (SamirTalwar-DA) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:432]: How does this work? `result` is no longer a `Try[…]`, as of #7179.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 2]
```diff
@@ -216,14 +221,17 @@ private[lf] case class PartialTransaction(
     }
 
   /** Finish building a transaction; i.e., try to extract a complete
-    *  transaction from the given 'PartialTransaction'. This fails if
-    *  the 'PartialTransaction' is not yet complete or has been
-    *  aborted.
+    *  transaction from the given 'PartialTransaction'. This returns:
+    * - a SubmittedTransaction in case of success ;
+    * - the 'PartialTransaction' itself if this later is not yet
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:226]: What do you mean by "later" here?

**Review by hurryabit — APPROVED** [review]
That change makes sense. Thank you.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:226]: I replaced "this later" by "it"

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala]: fixed.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 3]
```diff
@@ -233,11 +241,11 @@ private[lf] case class PartialTransaction(
           nodes
         )
         .fold(
-          s => Left(s"Cannot serialized the transaction: $s"),
-          tx => Right(Right(SubmittedTransaction(tx)))
+          s => SerializationError(s"Cannot serialized the transaction: $s"),
```
Comment (SamirTalwar-DA) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:244]: I suggest adding flavor text at the last possible moment.

```suggestion
          message => SerializationError(message),
```

(BTW, you also meant "serialize", not "serialized".)

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala [hunk 4]
```diff
@@ -427,9 +428,9 @@ class IdeClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClie
                     // Capture the result and exit.
                     result = results.toSeq
                 }
-              case Right(Left(x)) =>
-                throw new RuntimeException(s"Unexpected abort: $x")
-              case Left(msg) =>
+              case PartialTransaction.IncompleteTransaction(ptx) =>
+                throw new RuntimeException(s"Unexpected abort: $ptx")
+              case PartialTransaction.SerializationError(msg) =>
                 throw new RuntimeException(msg)
```
Comment (SamirTalwar-DA) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:434]:
```suggestion
                throw new RuntimeException(s"Cannot serialize the transaction: $msg")
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 5]
```diff
@@ -126,6 +126,11 @@ private[lf] object PartialTransaction {
     keys = Map.empty,
   )
 
+  sealed abstract class Result extends Product with Serializable
+  final case class CompleteTransaction(tx: SubmittedTransaction) extends Result
+  final case class IncompleteTransaction(ptx: PartialTransaction) extends Result
+  final case class SerializationError(msg: String) extends Result
```
Comment (SamirTalwar-DA) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:132]: If you make these two extend `RuntimeException` (and make `Result` a trait), you can just throw them directly.

```suggestion
  final case class IncompleteTransaction(ptx: PartialTransaction)
    extends RuntimeException(s"Unexpected abort: $ptx")
    with Result
  final case class SerializationError(msg: String)
    extends RuntimeException(s"Cannot serialize the transaction: $msg")
    with Result
```

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:132]: I definitively do not want to throw any of this. The important of `PartialTransaction#finish` method is in Engine where I have to handle explicitly all the 3 cases.

Comment (SamirTalwar-DA) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:132]: That's fair. We are throwing untyped `new RuntimeException`s elsewhere though; I'd appreciate it if, somehow, some more meaningful types emerged for those.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:434]: I think you missed this one.

**Review by SamirTalwar-DA — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:132]: Those untyped `new RuntimeException`s are part of dam-script. I prefer not touching too much this part. 
cc @cocreature

**Review by SamirTalwar-DA — APPROVED** [review]
Sold!

**Pull request merged**